src/query/binary.rs

```diff
@@ -270,57 +270,59 @@ impl std::iter::Iterator for BinaryEvaluatorVectorVector {
 
         // Only aligned in time vectors can be matched.
 
-        let (lv, rv) = loop {
-            let (lv, rv) = match (self.left.peek(), self.right.peek()) {
-                (Some(InstantVector(lv)), Some(InstantVector(rv))) => (lv, rv),
-                (None, _) | (_, None) => return None,
-                _ => unreachable!(),
-            };
+        let (lv, rv) = match (self.left.peek(), self.right.peek()) {
```

Comment:
Good catch! I think the code on the left has a bug. I'll double-check it tomorrow, but IIRC the idea was to break the loop only if the entries from both vectors are aligned in time or one of the iterators is exhausted. Otherwise, keep looping. So, the loop probably needs to stay, but the unconditional return at the bottom of the loop block should be reworked somehow.

Reply:
Oh, so that's the deal haha
So it'd be much appreciated if you could rewrite it (either on my branch or on a new one and Ill merge from master once you're done).
Its also a good idea to write a test that fails for the current implementation and passes after your fix, because I ran the tests after changing the implementation and they're all passing.

Keep us updated here please 😊